### PR TITLE
DPR Cluster Kubernetes

### PR DESCRIPTION
**Pull request by NikolaRaicevic**

### CODE REVIEW

File: cs/libdpr/benchmark/dprfinder/FASTER.benchmark.dprfinder/run.py [hunk 1]
```diff
@@ -17,5 +17,3 @@ def handle(self):
     host = socket.gethostbyname(socket.gethostname())
     server = socketserver.TCPServer((host, 15000), ScriptServer)
     server.serve_forever()
-
```
Comment (tli2) [inline cs/libdpr/benchmark/dprfinder/FASTER.benchmark.dprfinder/run.py:20]: nit: unintentional change?

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs [hunk 2]
```diff
@@ -12,20 +12,43 @@ namespace DprCounters
     public class CounterClient
     {
         private DprClient client;
-        private Dictionary<Worker, IPEndPoint> cluster;
-        private Thread backgroundThread;
+        private IPAddress ip;
+        Dictionary<Worker, EndPoint> cluster = new Dictionary<Worker, EndPoint>();
 
-        public CounterClient(IDprFinder dprFinder, Dictionary<Worker, IPEndPoint> cluster)
+        public CounterClient(string ip, int port)
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs]: nit: maybe remove this constructor because for other types of DprFinder this is not going to work. Alternatively, turn this into a factory method with the appropriate name (e.g., `static CounterClient CreateCounterClientForEnhancedDprFinder(string ip, int port)`)

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs [hunk 3]
```diff
@@ -12,20 +12,43 @@ namespace DprCounters
...
+            this.ip = IPAddress.Parse(ip);
+        }
+
+        public CounterClient(IDprFinder dprFinder)
+        {
+            // used when the Dpr Finder uses DNS for when the client is inside the cluster
+            // usually for testing purposes
             client = new DprClient(dprFinder);
-            this.cluster = cluster;
+            this.ip = null;
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs]: Is `ip` an optional member here then?  Note that you use ip later in `GetSession()` which is then used to create EndPoint for sessions. Maybe that'll lead to NPE if `ip` is null?

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs [hunk 4]
```diff
@@ -12,20 +12,43 @@ namespace DprCounters
...
+        }
+
+        public Dictionary<Worker, EndPoint> getCluster()
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs]: nit: `GetCluster`

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs [hunk 5]
```diff
@@ -12,20 +12,43 @@ namespace DprCounters
...
+        {
+            return cluster;
         }
         
         public CounterClientSession GetSession()
         {
+            var fetchedClusterInfo = client.FetchCluster();
+            foreach (var (key, value) in fetchedClusterInfo)
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs:33]: Does this ever remove any endpoints from `cluster` when nodes leave the cluster?

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs [hunk 6]
```diff
@@ -12,20 +12,43 @@ namespace DprCounters
...
+            {
+                cluster[key] = new IPEndPoint(ip, value.Item1);
+            }
             return new(client.GetSession(Guid.NewGuid()), cluster);
         }
 
+        public CounterClientSession GetSession(Dictionary<Worker, EndPoint> specifiedCluster)
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs]: I am guessing this is the band-aid from earlier due to the `ip` restriction? We should probably refactor this.

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClientSession.cs [hunk 7]
```diff
@@ -51,10 +51,10 @@ public long Increment(Worker worker, long amount, out long result)
                 header.Length + sizeof(long));
             header.CopyTo(new Span<byte>(serializationBuffer, sizeof(int), header.Length));
             BitConverter.TryWriteBytes(new Span<byte>(serializationBuffer, header.Length + sizeof(int), sizeof(long)), amount);
-            
+           
             // For simplicity, start a new socket every operation
             var endPoint = cluster[worker];
-            using var socket = new Socket(endPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/CounterClientSession.cs:57]: nit: why is this changed?

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClientSession.cs [hunk 8]
```diff
@@ -92,7 +91,7 @@ public bool Committed(long seq)
             
             var cp = versionTracker.GetCommitPoint();
             // Because the session is strictly sequential, operation will never be in exception list.
-            Debug.Assert(cp.ExcludedSerialNos.Count == 0);
+            // Debug.Assert(cp.ExcludedSerialNos.Count == 0);
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/CounterClientSession.cs:109]: nit: why is this changed?

File: cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinder.cs [hunk 9]
```diff
@@ -1,16 +1,54 @@
...
 using System.Diagnostics;
+using System.IO;
 using System.Net;
 using System.Net.Sockets;
 using System.Text;
 using System.Threading;
 
 namespace FASTER.libdpr
 {
+    public static class Extensions
```
Comment (tli2) [inline cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinder.cs]: Move this into `FASTER.libdpr.Utility` class?

File: cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinder.cs [hunk 10]
```diff
@@ -1,16 +1,54 @@
...
+        public static void LogBasic(string file, string text)
+        {
+            if(distributed)
+                Write(file, text);
+        }
+    }
     public class EnhancedDprFinder : IDprFinder
     {
-        private readonly Socket dprFinderConn;
+        private static readonly string basicLog = "/DprCounters/data/basic.txt";
```
Comment (tli2) [inline cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinder.cs:14]: nit: maybe try to avoid hardcoding log directory in this class. Have it in a centralized location that we can change when we run the program as opposed to a compile time constant

File: cs/libdpr/samples/DprCounters/DprCounters/DprCounters.csproj [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
     <PropertyGroup>
         <OutputType>Exe</OutputType>
-        <TargetFramework>netcoreapp3.1</TargetFramework>
+        <TargetFramework>net6.0</TargetFramework>
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/DprCounters.csproj:5]: nit: Is this intentional?

File: cs/libdpr/samples/DprCounters/DprCounters/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,20 @@
+FROM mcr.microsoft.com/dotnet/aspnet:6.0
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/Dockerfile:1]: Is this generated? Seems to have dpr counter hardcoded.

File: cs/libdpr/samples/DprCounters/DprCounters/Program.cs [hunk 13]
```diff
@@ -10,37 +11,40 @@ namespace DprCounters
 {
     class Program
     {
-        static void Main(string[] args)
+        static string DPR_FINDER_IP = "20.223.12.243"; // equal to $(minikube ip), which is the persistent IP of the DPR Finder
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/Program.cs]: nit: Best not to commit your instance specific IP

File: cs/libdpr/samples/DprCounters/DprCounters/CounterServer.cs [hunk 14]
```diff
@@ -12,6 +12,8 @@ namespace DprCounters
     /// </summary>
     public class CounterServer
     {
+        private static readonly string errorLog = "/DprCounters/data/errors.txt";
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/CounterServer.cs:15]: nit: try to have these in centralized locations that can be changed at runtime rather than compile-time constants at the class.

File: cs/libdpr/samples/DprCounters/DprCounters/CounterServer.cs [hunk 15]
```diff
@@ -24,7 +26,7 @@ public class CounterServer
         /// <param name="me"> id of worker in DPR cluster </param>
         /// <param name="checkpointDir"> directory name to write checkpoint files to </param>
         /// <param name="dprFinder"> DprFinder for the cluster </param>
-        public CounterServer(string ip, int port, Worker me, string checkpointDir, IDprFinder dprFinder)
+        public CounterServer(string ip, int port, WorkerInformation me, string checkpointDir, IDprFinder dprFinder)
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/CounterServer.cs:29]: Is WorkerInfo here specific to CounterServer or a more general notion in DPR cluster?

File: cs/libdpr/samples/DprCounters/DprCounters/Program.cs [hunk 16]
```diff
@@ -55,6 +59,99 @@ static void Main(string[] args)
...
+            var localDevice2 = new ManagedLocalStorageDevice("/DprCounters/data/dpr2.dat", deleteOnClose: true);
+            var localDeviceCluster1 = new ManagedLocalStorageDevice("/DprCounters/data/cluster1.dat", deleteOnClose: true);
+            var localDeviceCluster2 = new ManagedLocalStorageDevice("/DprCounters/data/cluster2.dat", deleteOnClose: true);
+            var device = new PingPongDevice(localDevice1, localDevice2);
+            var deviceCluster = new PingPongDevice(localDeviceCluster1, localDeviceCluster2);
+            EnhancedDprFinderServer backendServerFinder = new EnhancedDprFinderServer("0.0.0.0", 3000, new EnhancedDprFinderBackend(device, deviceCluster));
+            backendServerFinder.StartServer();
+        }
+
+        static void RunClientRight()
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/Program.cs]: nit: What does "Right" mean here? We probably want to be more descriptive and precise.

File: cs/libdpr/samples/DprCounters/DprCounters/Program.cs [hunk 17]
```diff
@@ -55,6 +59,99 @@ static void Main(string[] args)
...
+                string frontPort = Environment.GetEnvironmentVariable("FRONTEND_PORT");
+                RunCounterServer(DPR_FINDER_SERVICE, DPR_FINDER_PORT, guid, Int32.Parse(frontPort));
+                return;
+            }
+            if(args[0] == "backend")
+            {   
+                RunBackendServer();
+                return;
+            }
+            if(args[0] == "clientRight")
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/Program.cs]: nit: same as above

File: cs/libdpr/samples/DprCounters/DprCounters/python/start_cluster.py [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+import yaml
+
+
+class KubernetesCluster():
+
+    DEFAULT_STORAGE = "1Gi"
+    SUPPORTED_SERVER_TYPES = ["counter"]
+    BASE_PORT = 6380
+    BASE_AZURE_PORT = 31101
+    LOCAL_DIRECTORY = "/mnt/c/Users/cetko/OneDrive/Desktop/MEng Thesis/new_faster/cs/libdpr/samples/DprCounters/DprCounters"
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/python/start_cluster.py:32]: Can you put some of these into a config class that is taken in at initialization?

File: cs/libdpr/samples/DprCounters/DprCounters/python/start_cluster.py [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+    LOCAL_IMAGE = "cetko24/meng_project"
+    LOCAL_POLICY = "Never"
+    AZURE_DIRECTORY = "/home/nikola/FASTER/cs/libdpr/samples/DprCounters/DprCounters"
+    AZURE_IMAGE = "nikolameng.azurecr.io/meng"
+    AZURE_POLICY = "Always"
+
+
+    class Server():
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/python/start_cluster.py]: nit: does this need to be a nested class?

File: cs/libdpr/samples/DprCounters/DprCounters/python/start_cluster.py [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+        def __init__(self, type:str, storage_location:str = None, storage_capacity:str = None, cpu_request:str = None, cpu_limit:str = None, memory_request:str = None, memory_limit:str = None):
+            self.type = type
+            self.storage_location = storage_location
+            self.storage_capacity = storage_capacity
+            self.cpu_request = cpu_request
+            self.cpu_limit = cpu_limit
+            self.memory_request = memory_request
+            self.memory_limit = memory_limit
+
+    def __init__(self, checkpoint_dir:str = None, azure:bool = False) -> void:
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/python/start_cluster.py:39]: Let's try to have a bit more sophistication around Kubernetes backend --- as opposed to have boolean flags around azure, can we have some kind of polymorphism around k8s backend? (e.g., minikubes & azure, and the ability to easily add others like Amazon without needing to change code.)

File: cs/libdpr/samples/DprCounters/DprCounters/python/start_cluster.py [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+            k+=1
+
+    def killDprFinder(self) -> void:
+        self.core.delete_namespaced_pod(name="dpr-finder-0", namespace="default")
+
+    def stopDprFinder(self) -> void:
+        self.apps.delete_namespaced_stateful_set(name="dpr-finder", namespace = "default")
+        self.core.delete_namespaced_service(name="dpr-finder-svc", namespace="default")
+    
+    def startCounterServers(self, counterServerService:str = None, counterServiceStateful:str = None) -> void:
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/python/start_cluster.py:98]: Eventually we want to move away from only CounterServers and make it arbitrary state objects

File: cs/libdpr/samples/DprCounters/DprCounters/scripts/docker_stuff.sh [hunk 22]
```diff
@@ -0,0 +1,3 @@
+docker rmi cetko24/meng_project
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/scripts/docker_stuff.sh]: nit: what is "stuff" here?

File: cs/libdpr/samples/DprCounters/DprCounters/scripts/kubectl_stop.sh [hunk 23]
```diff
@@ -0,0 +1,10 @@
+kubectl delete svc counter-server-svc-0
```
Comment (tli2) [inline cs/libdpr/samples/DprCounters/DprCounters/scripts/kubectl_stop.sh:1]: Just to be clear: this is specific for a deployment and not a generic script right?

File: cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinder.cs [hunk 24]
```diff
@@ -22,13 +60,38 @@ public EnhancedDprFinder(Socket dprFinderConn)
...
-            dprFinderConn = new Socket(ipEndpoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+            this.ip = ip;
+            this.port = port;
+            ResetDprFinderConn();
+        }
+
+        private void ResetDprFinderConn()
+        {
+            EndPoint endpoint;
+            if(Char.IsDigit(ip[0]))
```
Comment (tli2) [inline cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinder.cs:40]: This sounds like a problem. What about `localhost`?

File: cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinderServer.cs [hunk 25]
```diff
@@ -22,8 +23,8 @@ namespace FASTER.libdpr
     /// </summary>
     public class EnhancedDprFinderServer : IDisposable
     {
+        private static readonly string debugLogFile = "/DprCounters/data/serverLog.txt";
```
Comment (tli2) [inline cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinderServer.cs:26]: nit: same as other hardcoded paths

File: cs/libdpr/src/FASTER.libdpr/dprfinder/resp/MessageUtil.cs [hunk 26]
```diff
@@ -1,16 +1,19 @@
...
+using System.IO;
+using System.Net;
 using System.Net.Sockets;
+using System.Text;
 
 namespace FASTER.libdpr
 {
     internal static class MessageUtil
     {
+        private static readonly string serverLog = "/DprCounters/data/server.txt";
```
Comment (tli2) [inline cs/libdpr/src/FASTER.libdpr/dprfinder/resp/MessageUtil.cs:13]: nit: same as other hardcoded paths

File: cs/libdpr/src/FASTER.libdpr/dprfinder/resp/RespUtil.cs [hunk 27]
```diff
@@ -1,11 +1,22 @@
 using System;
 using System.Collections.Generic;
+using System.Net;
 using System.Linq;
 
 namespace FASTER.libdpr
 {
     public static class RespUtil
     {
+        private static readonly Dictionary<string, int> typeToInt = new Dictionary<string, int>
```
Comment (tli2) [inline cs/libdpr/src/FASTER.libdpr/dprfinder/resp/RespUtil.cs:10]: Let's just use strings directly for cluster type unless there's a performance reason?

File: cs/libdpr/src/FASTER.libdpr/serverlib/SimpleVersionScheme.cs [hunk 28]
```diff
@@ -82,7 +82,6 @@ public bool TryAdvanceVersion(Action<long, long> criticalSection, long targetVer
             while (Interlocked.CompareExchange(ref versionChanged, ev, null) != null)
             {
             }
-
```
Comment (tli2) [inline cs/libdpr/src/FASTER.libdpr/serverlib/SimpleVersionScheme.cs]: nit: unintentional change?

File: cs/libdpr/samples/DprCounters/DprCounters/Program.cs (hunk 16 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/Program.cs]: Got rid of this. It meant "client is inside the cluster" but we don't need this at all and it is just confusing at this point

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs (hunk 2 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs]: Removed and made it so that the Counter Client takes in a DPR Finder but also an IP address of the cluster since the way we made it requires this IP. IP is also no longer an optional field so we are safe from warnings you mention below.

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs (hunk 6 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs]: Nah, this was a hack I was using from before we were able to FetchCluster(). I just gave it to the Session manually and pretended it was fetched. It had 0 references, removed.

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClientSession.cs (hunk 8 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/CounterClientSession.cs:109]: Back when I just got the code and started running it, this piece of code used to sometimes give me assertion errors so I commented it out. I commented it back in.

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClientSession.cs (hunk 7 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/CounterClientSession.cs:57]: Changed it back when we had the client use the DNS. Got rid of the DNS client so reverted this as well

File: cs/libdpr/samples/DprCounters/DprCounters/DprCounters.csproj (hunk 11 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/DprCounters.csproj:5]: Yes. It was much easier to get this to work with Docker and containerize it. Also seemed newer anyways and didn't cause any problems.

File: cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs (hunk 5 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/CounterClient.cs:33]: No, but once the DPR Finder has its view of the cluster updated, FetchCluster() will be returning a different cluster version. Same for addition. I added re-fetching the cluster when we are refreshing the DPR from the client. Will need to re-fetch for failed Increment() operations as well, but that comes later

File: cs/libdpr/samples/DprCounters/DprCounters/Dockerfile (hunk 12 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/Dockerfile:1]: This is not generated. This part of the process cannot be generated and has to look like this since it's codebase specific.  This file is implicitly used when building an image of the codebase with the docker_stuff script

File: cs/libdpr/samples/DprCounters/DprCounters/Program.cs (hunk 13 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/Program.cs]: Will remove for the next commit

File: cs/libdpr/samples/DprCounters/DprCounters/CounterServer.cs (hunk 15 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/CounterServer.cs:29]: A more general notion in the cluster. It is an expansion of Worker used heavily by the FetchCluster() operation and everything supporting that operation. It is used by the DPR Finder backend, DPR Finder, Client, Servers....

File: cs/libdpr/samples/DprCounters/DprCounters/scripts/docker_stuff.sh (hunk 22 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/scripts/docker_stuff.sh]: Renamed to a nicer name.

File: cs/libdpr/samples/DprCounters/DprCounters/scripts/kubectl_stop.sh (hunk 23 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/scripts/kubectl_stop.sh:1]: Yes, this script exists from way back when and I still use it because it is convenient. It is not actual "production" material. Eventually we will use the python code to safely dismantle the DPR Cluster.

File: cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinder.cs (hunk 24 above)
Comment (NikolaRaicevic) [reply cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinder.cs:40]: I am not sure what you mean here. Can you elaborate?

File: cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinderServer.cs (hunk 25 above)
Comment (NikolaRaicevic) [reply cs/libdpr/src/FASTER.libdpr/dprfinder/enhanced/EnhancedDprFinderServer.cs:26]: Resolving because it's a duplicate

File: cs/libdpr/src/FASTER.libdpr/dprfinder/resp/MessageUtil.cs (hunk 26 above)
Comment (NikolaRaicevic) [reply cs/libdpr/src/FASTER.libdpr/dprfinder/resp/MessageUtil.cs:13]: Resolving because it's a duplicate

File: cs/libdpr/samples/DprCounters/DprCounters/CounterServer.cs (hunk 14 above)
Comment (NikolaRaicevic) [reply cs/libdpr/samples/DprCounters/DprCounters/CounterServer.cs:15]: Resolving because it's a duplicate

File: cs/libdpr/src/FASTER.libdpr/dprfinder/resp/RespUtil.cs (hunk 27 above)
Comment (NikolaRaicevic) [reply cs/libdpr/src/FASTER.libdpr/dprfinder/resp/RespUtil.cs:10]: Added a TODO for this. I agree that we should do this, and I will do this later, but I don't want to deal with the whole communication setup at the moment. The only reason I chose an int is because it has a constant size and it was more straightforward to send it that way.
